eth/downloader/statesync.go

```diff
@@ -210,6 +207,42 @@ func (d *Downloader) runStateSync(s *stateSync) *stateSync {
 	}
 }
 
+// spindownStateSync 'drains' the outstanding requests; some will be delivered and other
+// will time out. This is to ensure that when the next stateSync starts working, all peers
+// are marked as idle and de facto _are_ idle.
+func (d *Downloader) spindownStateSync(active map[string]*stateReq, timeout chan *stateReq, peerDrop chan *peerConnection, next *stateSync) *stateSync {
```

Comment:
For all sent requests, we can classify them into these categories:

- Sent and deliveries the response
- Sent but timeout
- Sent but the peer is dropped
- Sent and still fetching 

For the first three categories, the result(completed or failed) will be cached in the `finished`. And these relevant peers still be marked as BUSY.

For the last category, we will wait for them until they are (a) timeout (b) delivery (c) peer dropped and then mark them as idle.

So the issue here is we don't handle the first three categories properly which will eventually lead to we can never reuse these peers again.

Reply:
In the spin-down, whenever
- An active peers delivers, or 
- An active peer is dropped, or 
- A request times out, then

The peer it concerns is marked as idle.

Reply:
Very nice catch, thanks a lot @rjl493456442 ! I've pushed a fix